compiler/rustc_lint/src/reachable_patterns.rs

```diff
@@ -0,0 +1,152 @@
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn check_pat(&mut self, cx: &LateContext<'tcx>, pat: &'tcx Pat<'_>) {
+        if let PatKind::Struct(_, fields @ [_, ..], true) = &pat.kind {
+            if let Some(ty::Adt(def, _)) = cx.maybe_typeck_results().map(|r| r.pat_ty(pat).kind()) {
+                if let Some(var) = def.variants.iter().next() {
+                    if var.is_field_list_non_exhaustive() {
+                        let missing = get_missing_fields(fields, &var.fields);
+                        if !missing.is_empty() {
+                            let last = fields.last().unwrap().span;
+                            lint(cx, pat.span, last, missing)
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+fn lint(cx: &LateContext<'_>, all: Span, _last: Span, missing: Vec<String>) {
+    cx.struct_span_lint(REACHABLE_PATTERNS, all, |lint: LintDiagnosticBuilder<'_>| {
+        let mut l = lint.build("missing patterns of non_exhaustive type");
+        l.help(&format!("add {} to match all reachable patterns", missing.join(",")));
+        l.emit();
+    });
+}
+
+fn get_missing_variants<'tcx>(
+    cx: &LateContext<'tcx>,
+    arms: &[Arm<'_>],
+    e: &'tcx Expr<'_>,
+) -> Vec<String> {
+    let ty = cx.typeck_results().expr_ty(e);
+    let mut missing_variants = vec![];
+    if let ty::Adt(def, _) = ty.kind() {
+        for variant in &def.variants {
+            missing_variants.push(variant);
+        }
+    }
+    let mut has_attr = false;
+    for arm in arms {
+        has_attr |= has_reachable_attr(cx, arm);
+
+        if let PatKind::Path(ref path) = arm.pat.kind {
+            if let QPath::Resolved(_, p) = path {
+                missing_variants.retain(|e| e.ctor_def_id != Some(p.res.def_id()));
+            }
+        } else if let PatKind::TupleStruct(QPath::Resolved(_, p), ref patterns, ..) = arm.pat.kind {
+            let is_pattern_exhaustive =
+                |pat: &&Pat<'_>| matches!(pat.kind, PatKind::Wild | PatKind::Binding(.., None));
+            if patterns.iter().all(is_pattern_exhaustive) {
+                missing_variants.retain(|e| e.ctor_def_id != Some(p.res.def_id()));
+            }
+        }
+    }
+
+    if !has_attr {
+        return vec![];
+    }
+
+    let missing_variants =
+        missing_variants.iter().map(|v| format!("`{}`", cx.tcx.def_path_str(v.def_id))).collect();
+    missing_variants
+}
+
+fn get_missing_fields(used: &[PatField<'_>], defined: &[ty::FieldDef]) -> Vec<String> {
+    let mut missing = vec![];
+
+    for def in defined {
+        if !used.iter().any(|f| f.ident == def.ident) && def.vis.is_visible_locally() {
+            missing.push(format!("`{}`", def.ident.name))
+        }
+    }
+
+    missing
+}
+
+fn has_reachable_attr(cx: &LateContext<'_>, arm: &Arm<'_>) -> bool {
+    cx.tcx.hir().attrs(arm.hir_id).iter().any(|attr: &Attribute| {
```

Comment:
To the best of my knowledge, most lints don't actually have to check for an attribute like this - the logic that emits lints determines the current lint level from the `Span` and emits the lint if appropriate given that lint level. 

As the default for this lint is `Allow`, emitting it unconditionally when you see a non-exhaustive struct/enum/enum variant will be silent unless that statement is annotated with a deny or warn attribute for that lint. It looks like this is adding both `{allow,deny,warn}(reachable)` and `{allow,deny,warn}(reachable_patterns)` and I don't think any of our other lints work like that.